README.md
# San Andreas RadioEx
Copyright © Thiekus (Faris Khowarizmi) 2014 - 2020
Free and Open Source under MPL 2.0
Project repository at https://github.com/thiekus/sa-radioex

San Andreas RadioEx is asi script that allow you to replace current radio stations.
It's like San Andreas Own Radio Station by HackMan128, but with lot improvements:
* It's asi script! Just run gta-sa.exe and it will run.
* No need to replace any GTA files! It has own stream file emulation.
* It have cool loading sound when stream buffer on progress
* It have custom notification sound when cannot open stream (English and Indonesian).
* Resolves most anticipated bug in SAORS like missing sound.
* You can replace the radio name without replacing on *.gxt file (altrough have some limitations).
* Not only MP3 radio streams, RadioEx also support MP4 and Opus.
* Free and open source under LGPL 2.1 (coded in Delphi).

How to Install?

I recomended to use Slinet's ASI Loader (It's performs better):
http://www.gtagarage.com/mods/show.php?id=21709

Then, place RadioEx.asi, RadioEx.ini and some bass library support on GTA SA directory. Edit RadioEx.ini with your mind! For list of Radio Streaming URL, just Google it :)

(IMPORTANT! You must set EnableBassInit to 0 if you have CLEO installed or asi that use bass.dll dependency) And, if you indonesian, leave NoticeLang to ID, otherwise must EN.
(To hear "Cannot open radio baby!", instead "Tidak bisa membuka radio, gan!")

Play and feel the difference!

Changelogs:

### V 1.1.0.253 (23 August 2014)
````
+ Faster radio change.
+ Adverts stream are emulated.
````

### V 1.0.2.244 (21 August 2014)
````
* Fix string memory buffer leaks and c call function fault. This will resolve problem when creating thread but insufficient memory.
* Some minor changes on initialization stage.
* Change from WinAPI CreateThread to BeginThread.
````

### V 1.0.1.234
````
* Initial release
````

Known issues and limitations:
* It only support GTA SA version 1.0
* It's have RadioNamePatch feature, sometimes like just not work (will back to original, but instantly patched) because I not patch all possible text loader for technical reasons.
* Only internet radio streaming player ability, your music files will play soon... :)

## License

San Andreas RadioEx is licensed under Mozilla Public License (MPL) 2.0.


codehook/CHANGES.txt
2006-08-08: Version 1.3c
- Small fix in "CodeLen.pas" (Delphi 5 only).

2006-01-01: Version 1.5
- Some small changes to avoid warnings under Delphi 2006.

2005-10-23: Version 1.4
- Almost complete rewrite of CodeLen.pas with the new functions
  "AnalyzeCpuInstructionSequence", "LengthOfCpuInstructionSequence", and
  "CopyCpuInstructionSequence".
- Modified "CreateGenericCodeHook" to use two different branch types
  instead of a pc-relative displacement.
- Removed the flag "Is16Bit" - did never work for 16-bit code.
- Still no support for Windows 95/98/Me system area ($80000000-$FFFFFFFF).

Version 1.3 (non-public)
- Removed from WebSite because of several problems regarding relocation of
  relative displacements, and with Windows 9x/Me and global hooks.
- Added "AnalyzeCpuInstruction" for a thorough analysis of the instruction,
  including referenced code and data memory.

2005-09-16: Version 1.2
- First public version (added sample application).
- Enhanced "PatchMemory" by first overwriting the first byte with $CC.
- Added some AMD specific instructions and 3DNow to "CodeLen.pas".

2005-09-15: Version 1.1
- Added module "CodeMem.pas" for extensible code memory allocation.
- Wrote optimized code memory manager "CodeMemOpt.pas".
- Enhanced "PatchMemory" by "FlushInstructionCache".
- Added function "IsJumpInstruction" to "CodeLen.pas".

2005-09-15: Version 1.0
- Initial version, derived from the "PrintToFile" project.


bass/bassopus.pas
{
  BASSOPUS 2.4 Delphi unit
  Copyright (c) 2012 Un4seen Developments Ltd.

  See the BASSOPUS.CHM file for more detailed documentation
}

unit BassOPUS;

interface

{$IFDEF MSWINDOWS}
uses BASS, Windows;
{$ELSE}
uses BASS;
{$ENDIF}

const
  // BASS_CHANNELINFO type
  BASS_CTYPE_STREAM_OPUS        = $11200;

  // Additional attributes
  BASS_ATTRIB_OPUS_ORIGFREQ     = $13000;

const
{$IFDEF MSWINDOWS}
  bassopusdll = 'bassopus.dll';
{$ENDIF}
{$IFDEF LINUX}
  bassopusdll = 'libbassopus.so';
{$ENDIF}
{$IFDEF MACOS}
  bassopusdll = 'libbassopus.dylib';
{$ENDIF}

function BASS_OPUS_StreamCreateFile(mem:BOOL; fl:pointer; offset,length:QWORD; flags:DWORD): HSTREAM; {$IFDEF MSWINDOWS}stdcall{$ELSE}cdecl{$ENDIF}; external bassopusdll;
function BASS_OPUS_StreamCreateURL(url: PAnsiChar; offset: DWORD; flags: DWORD; proc: DOWNLOADPROC; user: Pointer): HSTREAM; {$IFDEF MSWINDOWS}stdcall{$ELSE}cdecl{$ENDIF}; external bassopusdll;
function BASS_OPUS_StreamCreateFileUser(system,flags:DWORD; var procs:BASS_FILEPROCS; user:Pointer): HSTREAM; {$IFDEF MSWINDOWS}stdcall{$ELSE}cdecl{$ENDIF}; external bassopusdll;

implementation

end.

codehook/CodeMemOpt.pas
{
  CodeMemOpt.pas

  Delphi unit containing a small memory management for virtually allocated
  memory that allows execution of run-time created code. This way many small
  fragments share one 4K VirtualAlloc page.

  Note: this unit is not optimized for speed (no necessity).

  Version 1.4 - Always find the most current version at
  http://flocke.vssd.de/prog/code/pascal/codehook/

  Copyright (C) 2005, 2006 Volker Siebert <flocke@vssd.de>
  All rights reserved.

  Permission is hereby granted, free of charge, to any person obtaining a
  copy of this software and associated documentation files (the "Software"),
  to deal in the Software without restriction, including without limitation
  the rights to use, copy, modify, merge, publish, distribute, sublicense,
  and/or sell copies of the Software, and to permit persons to whom the
  Software is furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
  DEALINGS IN THE SOFTWARE.
}

unit CodeMemOpt;

interface

uses
  Windows, SysUtils, CodeMem;

function OptGetCodeMem(Size: integer): pointer;
function OptFreeCodeMem(P: pointer): integer;

implementation

const
  CHUNK_ROUND_SIZE = 16;
  CHUNK_USED_BIT   = $40000000;
  CHUNK_SIZE_MASK  = $3FFFFFFF;

  PAGE_MAGIC       = $6D4D7356;

type
  // The header of a complete page
  PPageHeader = ^TPageHeader;
  TPageHeader = packed record
    Magic: cardinal;            // PAGE_MAGIC
    Next: PPageHeader;          // pointer to the next page header
    Size: integer;              // size of this page
    Dummy: integer;             // not used
  end;

  // A single chunk of allocated memory
  PPageChunk = ^TPageChunk;
  TPageChunk = packed record
    Base: PPageHeader;          // the page's header
    Size: integer;              // complete size incl. this header
  end;

var
  PageCrit: TRTLCriticalSection;        // Critical section to be thread-enabled
  PageCritFlag: boolean = false;        // Flag if InitializeCriticalSection was done
  PageSize: integer = 0;                // System VirtualAlloc page size
  PageList: PPageHeader = nil;          // List of allocated pages

{ The optimized code memory manager }

var
  OptCodeMemManager: TCodeMemoryManager = (
    GetMem:     OptGetCodeMem;
    FreeMem:    OptFreeCodeMem;
    ReallocMem: SysReallocMem
  );

function OptGetCodeMem(Size: integer): pointer;
var
  Page: PPageHeader;

  { Append a new page to the list of pages
  }
  function AppendNewPage(ForceSize: integer = 0): PPageHeader;
  begin
    // Round to full page size
    if ForceSize = 0 then
      ForceSize := PageSize
    else
      ForceSize := PageSize * ((ForceSize + PageSize - 1) div PageSize);

    // Get new physical page
    Result := VirtualAlloc(nil, ForceSize, MEM_COMMIT, PAGE_EXECUTE_READWRITE);
    if Result = nil then
      exit;

    with PPageChunk(integer(Result) + SizeOf(TPageHeader))^ do
    begin
      Size := ForceSize - SizeOf(TPageHeader);
      Base := Result;
    end;

    with Result^ do
    begin
      Magic := PAGE_MAGIC;
      Size := ForceSize;
      Dummy := 0;
      Next := PageList;
    end;

    PageList := Result;
  end;

  { Initialize the page size and the page list
  }
  procedure InitPages;
  var
    Info: TSystemInfo;
  begin
    GetSystemInfo(Info);
    PageSize := Info.dwPageSize;
    if PageSize < 4096 then
      PageSize := 4096;

    if PageSize mod CHUNK_ROUND_SIZE <> 0 then
      PageSize := PageSize - PageSize mod CHUNK_ROUND_SIZE;
  end;

  { Allocate a piece of memory in the given page. Size includes the chunk
    header overhead and is already rounded to CHUNK_ROUND_SIZE.
  }
  function AllocInPage(Page: PPageHeader; Size: integer): pointer;
  var
    Chunk, Next, Limit: integer;
  begin
    Result := nil;

    Chunk := integer(Page) + SizeOf(TPageHeader);
    Limit := integer(Page) + Page^.Size;

    while (Chunk < Limit) and (Result = nil) do
    begin
      if (PPageChunk(Chunk)^.Size and CHUNK_USED_BIT) <> 0 then
        inc(Chunk, PPageChunk(Chunk)^.Size and CHUNK_SIZE_MASK)
      else
      begin
        Next := Chunk + PPageChunk(Chunk)^.Size;
        while (Next < Limit) and ((PPageChunk(Next)^.Size and CHUNK_USED_BIT) = 0) do
        begin
          inc(PPageChunk(Chunk)^.Size, PPageChunk(Next)^.Size);
          Next := Chunk + PPageChunk(Chunk)^.Size;
        end;

        if PPageChunk(Chunk)^.Size > Size then
        begin
          // Allocate from the end, so the next alloc
          // finds the free block as fast as we did.
          dec(PPageChunk(Chunk)^.Size, Size);
          Next := Chunk + PPageChunk(Chunk)^.Size;
          PPageChunk(Next)^.Base := Page;
          PPageChunk(Next)^.Size := Size or CHUNK_USED_BIT;
          Result := pointer(Next + SizeOf(TPageChunk));
        end
        else if PPageChunk(Chunk)^.Size = Size then
        begin
          // Fits exactly
          PPageChunk(Chunk)^.Size := Size or CHUNK_USED_BIT;
          Result := pointer(Chunk + SizeOf(TPageChunk));
        end;

        Chunk := Next;
      end;
    end;
  end;

  { Move the given page to the top of the page list
  }
  procedure LastRecentlyUsed(Page: PPageHeader);
  var
    Fix: ^PPageHeader;
  begin
    Fix := @PageList;
    while Fix^ <> Page do
      Fix := @(Fix^)^.Next;

    Fix^ := Page^.Next;
    Page^.Next := PageList;
    PageList := Page;
  end;

begin
  if not PageCritFlag then
  begin
    PageCritFlag := true;
    InitializeCriticalSection(PageCrit);
  end;

  EnterCriticalSection(PageCrit);
  try
    if PageSize = 0 then
      InitPages;

    Result := nil;
    Size := (Size + SizeOf(TPageChunk) + CHUNK_ROUND_SIZE - 1) and (-CHUNK_ROUND_SIZE);

    Page := PageList;
    while Page <> nil do
    begin
      Result := AllocInPage(Page, Size);
      if Result <> nil then
        break;

      Page := Page^.Next;
    end;

    if Result = nil then
    begin
      if Size > PageSize - 128 then
        Page := AppendNewPage(Size + SizeOf(TPageHeader))
      else
        Page := AppendNewPage(0);

      Result := AllocInPage(Page, Size);
    end;

    if Page <> PageList then
      if 8 * Size <= PageSize then
        LastRecentlyUsed(Page);

    if Result = nil then
      SetLastError(ERROR_NOT_ENOUGH_MEMORY);
  finally
    LeaveCriticalSection(PageCrit);
  end;
end;

{ Free the memory pointed to by P
}
function OptFreeCodeMem(P: pointer): integer;
var
  Chunk: PPageChunk;
begin
  Result := 0;

  if P = nil then
    exit;

  if (not PageCritFlag) or (PageSize = 0) or (PageList = nil) then
  begin
    SetLastError(ERROR_INVALID_FUNCTION);
    Result := 1;
    exit;
  end;

  EnterCriticalSection(PageCrit);
  try
    if P <> nil then
    begin
      Chunk := PPageChunk(integer(P) - SizeOf(TPageChunk));
      // Check the chunk and the page
      if ((Chunk^.Size and CHUNK_USED_BIT) = 0) or
         ((Chunk^.Size and (CHUNK_ROUND_SIZE - 1)) <> 0) or
         (Chunk^.Base^.Magic <> PAGE_MAGIC) or
         (cardinal(Chunk) < cardinal(Chunk^.Base)) or
         (cardinal(Chunk) >= cardinal(Chunk^.Base) + cardinal(Chunk^.Base^.Size)) then
      begin
        SetLastError(ERROR_INVALID_BLOCK);
        Result := 1;
      end
      else
        // Everything seems to be ok
        Chunk^.Size := Chunk^.Size and CHUNK_SIZE_MASK;
    end;
  finally
    LeaveCriticalSection(PageCrit);
  end;
end;

initialization
  SetCodeMemManager(OptCodeMemManager);
finalization
  if PageCritFlag then
  begin
    DeleteCriticalSection(PageCrit);
    PageCritFlag := false;
  end;
end.



codehook/Sample/fmain.pas
{
  fmain.pas

  This file is part of the GenCodeHook.pas sample application.
  Info at http://flocke.vssd.de/prog/code/pascal/codehook/

  Copyright (C) 2005, 2006 Volker Siebert <flocke@vssd.de>
  All rights reserved.
}

unit fmain;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, ComCtrls, ExtCtrls, StdCtrls;

type
  TForm1 = class(TForm)
    Shape1: TShape;
    Shape2: TShape;
    Label1: TLabel;
    Edit1: TEdit;
    Button1: TButton;
    Label2: TLabel;
    Edit2: TEdit;
    Button2: TButton;
    Bevel1: TBevel;
    Button3: TButton;
    OpenDialog1: TOpenDialog;
    SaveDialog1: TSaveDialog;
    Label3: TLabel;
    Label4: TLabel;
    Timer1: TTimer;
    procedure FormClose(Sender: TObject; var Action: TCloseAction);
    procedure Timer1Timer(Sender: TObject);
    procedure FormHide(Sender: TObject);
    procedure FormShow(Sender: TObject);
    procedure Button3Click(Sender: TObject);
    procedure Button2Click(Sender: TObject);
    procedure Button1Click(Sender: TObject);
  private
    { Private-Deklarationen }
  public
    { Public-Deklarationen }
    FShapes: array [0 .. 1] of TShape;
    FFlashPoint: array [0 .. 1] of cardinal;
  end;

var
  Form1: TForm1;

implementation

{$R *.dfm}

{----------------------------------------------------------------------}

uses
  GenCodeHook, CodeMemOpt;

const
  CFlashDurance = 5;

type
  NTSTATUS = LongInt;
  TFnNtReadFile = function(FileHandle, Event: THandle; ApcRoutine: Pointer;
    ApcContext: Pointer; IoStatusBlock: Pointer; Buffer: Pointer;
    Length: ULONG; ByteOffset: Pointer; Key: Pointer): NTSTATUS; stdcall;
  TFnNtWriteFile = function(FileHandle, Event: THandle; ApcRoutine: Pointer;
    ApcContext: Pointer; IoStatusBlock: Pointer; Buffer: Pointer;
    Length: ULONG; ByteOffset: Pointer; Key: Pointer): NTSTATUS; stdcall;
  TFnNtCreateFile = function(FileHandle: Pointer; DesiredAccess: ACCESS_MASK;
    ObjectAttributes: Pointer; IoStatusBlock: Pointer;
    AllocationSize: Pointer; FileAttributes: ULONG;
    ShareAccess: ULONG; CreateDisposition: ULONG; CreateOptions: ULONG;
    EaBuffer: Pointer; EaLength: ULONG): NTSTATUS; stdcall;

  TFnReadFile = function(hFile: THandle; var Buffer; nNumberOfBytesToRead: DWORD;
    var lpNumberOfBytesRead: DWORD; lpOverlapped: POverlapped): BOOL; stdcall;
  TFnWriteFile = function(hFile: THandle; const Buffer; nNumberOfBytesToWrite: DWORD;
    var lpNumberOfBytesWritten: DWORD; lpOverlapped: POverlapped): BOOL; stdcall;

var
  OldNtReadFile: TFnNtReadFile;
  OldNtWriteFile: TFnNtWriteFile;
  OldNtCreateFile: TFnNtCreateFile;
  OldReadFile: TFnReadFile;
  OldWriteFile: TFnWriteFile;

const
  CSlotColor: array [0 .. 1] of TColor = ( clGreen, clRed );

procedure DoBefore(Slot: integer);
begin
  if GetCurrentThreadID = MainThreadID then
  begin
    Form1.FFlashPoint[Slot] := GetTickCount + CFlashDurance;

    if Form1.FShapes[Slot].Brush.Color <> CSlotColor[Slot] then
    begin
      Form1.FShapes[Slot].Brush.Color := CSlotColor[Slot];
      Form1.FShapes[Slot].Repaint;
    end;
  end;
end;

procedure DoAfter(Slot: integer);
begin
  if GetCurrentThreadID = MainThreadID then
    Application.ProcessMessages;
end;

function NewNtReadFile(FileHandle, Event: THandle; ApcRoutine: Pointer;
  ApcContext: Pointer; IoStatusBlock: Pointer; Buffer: Pointer;
  Length: ULONG; ByteOffset: Pointer; Key: Pointer): NTSTATUS; stdcall;
begin
  DoBefore(0);
  Result := OldNtReadFile(FileHandle, Event, ApcRoutine, ApcContext,
              IoStatusBlock, Buffer, Length, ByteOffset, Key);
  DoAfter(0);
end;

function NewNtWriteFile(FileHandle, Event: THandle; ApcRoutine: Pointer;
  ApcContext: Pointer; IoStatusBlock: Pointer; Buffer: Pointer;
  Length: ULONG; ByteOffset: Pointer; Key: Pointer): NTSTATUS; stdcall;
begin
  DoBefore(1);
  Result := OldNtWriteFile(FileHandle, Event, ApcRoutine, ApcContext,
              IoStatusBlock, Buffer, Length, ByteOffset, Key);
  DoAfter(1);
end;

function NewNtCreateFile(FileHandle: Pointer; DesiredAccess: ACCESS_MASK;
  ObjectAttributes: Pointer; IoStatusBlock: Pointer;
  AllocationSize: Pointer; FileAttributes: ULONG;
  ShareAccess: ULONG; CreateDisposition: ULONG; CreateOptions: ULONG;
  EaBuffer: Pointer; EaLength: ULONG): NTSTATUS; stdcall;
const
  FILE_WRITE_DATA  = $0002;
  FILE_APPEND_DATA = $0004;
var
  Slot: integer;
begin
  if (DesiredAccess and (FILE_WRITE_DATA or FILE_APPEND_DATA)) <> 0 then
    Slot := 1
  else
    Slot := 0;

  DoBefore(Slot);
  Result := OldNtCreateFile(FileHandle, DesiredAccess, ObjectAttributes,
              IoStatusBlock, AllocationSize, FileAttributes, ShareAccess,
              CreateDisposition, CreateOptions, EaBuffer, EaLength);
  DoAfter(Slot);
end;

function NewReadFile(hFile: THandle; var Buffer; nNumberOfBytesToRead: DWORD;
  var lpNumberOfBytesRead: DWORD; lpOverlapped: POverlapped): BOOL; stdcall;
begin
  DoBefore(0);
  Result := OldReadFile(hFile, Buffer, nNumberOfBytesToRead, lpNumberOfBytesRead,
              lpOverlapped);
  DoAfter(0);
end;

function NewWriteFile(hFile: THandle; const Buffer; nNumberOfBytesToWrite: DWORD;
  var lpNumberOfBytesWritten: DWORD; lpOverlapped: POverlapped): BOOL; stdcall;
begin
  DoBefore(1);
  Result := OldWriteFile(hFile, Buffer, nNumberOfBytesToWrite,
              lpNumberOfBytesWritten, lpOverlapped);
  DoAfter(1);
end;

procedure UninstallPatch;
begin
  RemoveGenericCodeHook(@OldNtReadFile);
  RemoveGenericCodeHook(@OldNtWriteFile);
  RemoveGenericCodeHook(@OldNtCreateFile);
  RemoveGenericCodeHook(@OldReadFile);
  RemoveGenericCodeHook(@OldWriteFile);
end;

function InstallPatch: boolean;
var
  Module: HMODULE;
  ok: boolean;

  function GPA(Name: PChar): Pointer;
  begin
    Result := GetProcAddress(Module, Name);
  end;

begin
  Module := GetModuleHandle('ntdll.dll');
  if Module <> 0 then
  begin
    ok := CreateGenericCodeHook(GPA('NtReadFile'),   @NewNtReadFile,   @OldNtReadFile) and
          CreateGenericCodeHook(GPA('NtWriteFile'),  @NewNtWriteFile,  @OldNtWriteFile) and
          CreateGenericCodeHook(GPA('NtCreateFile'), @NewNtCreateFile, @OldNtCreateFile);
  end
  else
  begin
    Module := GetModuleHandle('kernel32.dll');
    ok := CreateGenericCodeHook(GPA('ReadFile'),  @NewReadFile,  @OldReadFile) and
          CreateGenericCodeHook(GPA('WriteFile'), @NewWriteFile, @OldWriteFile);
  end;

  if not ok then
  begin
    UninstallPatch;
    MessageDlg('CreateGenericCodeHook did not work!', mtError, [mbOk], 0);
  end;

  Result := ok;
end;

{----------------------------------------------------------------------}

procedure TForm1.FormShow(Sender: TObject);
begin
  FShapes[0] := Shape1;
  FShapes[1] := Shape2;

  FFlashPoint[0] := 0;
  FFlashPoint[1] := 0;

  if not InstallPatch then
    Close;
end;

procedure TForm1.FormHide(Sender: TObject);
begin
  UninstallPatch;
end;

procedure TForm1.FormClose(Sender: TObject; var Action: TCloseAction);
begin
  UninstallPatch;
end;

procedure TForm1.Button1Click(Sender: TObject);
begin
  OpenDialog1.FileName := Edit1.Text;
  if OpenDialog1.Execute then
    Edit1.Text := OpenDialog1.FileName;
end;

procedure TForm1.Button2Click(Sender: TObject);
begin
  SaveDialog1.FileName := Edit2.Text;
  if SaveDialog1.Execute then
    Edit2.Text := SaveDialog1.FileName;
end;

procedure TForm1.Button3Click(Sender: TObject);
begin
  CopyFile(PChar(Edit1.Text), PChar(Edit2.Text), false);
  MessageBeep(MB_ICONINFORMATION);
end;

procedure TForm1.Timer1Timer(Sender: TObject);
var
  k: cardinal;
begin
  k := GetTickCount;

  if (FFlashPoint[0] <> 0) and (k > FFlashPoint[0]) then
  begin
    FFlashPoint[0] := 0;
    Shape1.Brush.Color := clWhite;
    Shape1.Repaint;
  end;

  if (FFlashPoint[1] <> 0) and (k > FFlashPoint[1]) then
  begin
    FFlashPoint[1] := 0;
    Shape2.Brush.Color := clWhite;
    Shape2.Repaint;
  end;
end;

end.


codehook/GenCodeHook.pas
{
  GenCodeHook.pas

  Delphi unit containing functions to do a generic code hook by replacing the
  code location to be patched by a jump to the new location. Additionally, a
  new code fragment is created that allows to call the old function.

  Version 1.5 - Always find the most current version at
  http://flocke.vssd.de/prog/code/pascal/codehook/

  Copyright (C) 2005, 2006 Volker Siebert <flocke@vssd.de>
  All rights reserved.

  Permission is hereby granted, free of charge, to any person obtaining a
  copy of this software and associated documentation files (the "Software"),
  to deal in the Software without restriction, including without limitation
  the rights to use, copy, modify, merge, publish, distribute, sublicense,
  and/or sell copies of the Software, and to permit persons to whom the
  Software is furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
  DEALINGS IN THE SOFTWARE.
}

unit GenCodeHook;

interface

uses
  Windows, SysUtils;

{ Creates a generic code hook by patching "Proc" to execute "NewProc" instead.
  Before patching, "OldProc" receives a pointer to a function that can be used
  to call the old code (though not at the same location).
}
function CreateGenericCodeHook(Proc, NewProc: pointer; var OldProc: pointer): boolean;

{ Restore a patched location and free the memory
}
procedure RemoveGenericCodeHook(var OldProc: pointer);

{ Function that patches protected memory
}
function PatchMemory(Where: pointer; const What; Size: integer;
  PadSize: integer = 0; PadByte: byte = $90): boolean;

implementation

uses
  CodeLen, CodeMem;

{$DEFINE ABSOLUTE_ADDRESSES}
{.$DEFINE ABS_JUMP_INDIRECT}

type
  TNearJmp = packed record
    Insn: byte;         // $E9 = jmp disp32
    Disp: integer;
  end;

  TAbsJmp = packed record
{$IFDEF ABS_JUMP_INDIRECT}
    Insn1: byte;        // $FF
    Insn2: byte;        // $25 (%00-100-101) = jmp dword ptr [offs]
    Addr: pointer;
{$ELSE}
    Insn1: byte;        // $68 = push imm32
    Addr: pointer;
    Insn2: byte;        // $C3 = ret
{$ENDIF}
  end;

const
  SIGNATURE   = $68636746;  // 'Fgch'
{$IFDEF ABSOLUTE_ADDRESSES}
  PATCH_SIZE  = SizeOf(TAbsJmp);
{$ELSE}
  PATCH_SIZE  = SizeOf(TNearJmp);
{$ENDIF}
  BACKUP_SIZE = MAX_IA32_INSTRUCTION_LENGTH + PATCH_SIZE;

type
  TGenCodeHookHeader = record
    BupLen: integer;
    Backup: array [0 .. BACKUP_SIZE - 1] of byte;
    Location: pointer;
{$IFDEF ABS_JUMP_INDIRECT}
    JumpHere: pointer;
{$ENDIF}
    Signature: cardinal;
  end;

  PGenCodeHookStruct = ^TGenCodeHookStruct;
  TGenCodeHookStruct = record
    Header: TGenCodeHookHeader;
    Code: array [0 .. SizeOf(TNearJmp) - 1] of byte;  // dynamisch
  end;

{ Pointer increment
}
function pinc(Ptr: pointer; incr: integer = 1): pointer;
begin
  Result := pointer(integer(Ptr) + incr);
end;

{ Function that patches protected memory
}
function PatchMemory(Where: pointer; const What; Size: integer;
  PadSize: integer = 0; PadByte: byte = $90): boolean;
var
  OldProtect: cardinal;
begin
  if Size <= 0 then
    Result := true
  else
  begin
    Result := VirtualProtect(Where, Size, PAGE_EXECUTE_READWRITE, OldProtect);
    if Result then
      try
        if Size = 4 then
          PCardinal(Where)^ := PCardinal(@What)^
        else if Size = 2 then
          PWord(Where)^ := PWord(@What)^
        else if Size = 1 then
          PByte(Where)^ := PByte(@What)^
        else
        begin
          PByte(Where)^ := $CC;   // int3
          Move(pinc(@What)^, pinc(Where)^, Size - 1);
          PByte(Where)^ := PByte(@What)^;
        end;

        if Size < PadSize then
        begin
          FillChar(pinc(Where, Size)^, PadSize - Size, PadByte);
          Size := PadSize;
        end;

        FlushInstructionCache(GetCurrentProcess(), Where, Size);
      finally
        VirtualProtect(Where, Size, OldProtect, OldProtect);
      end;
  end;
end;

{ Calculate the displacement from Source to Target
  (Intel IA-32 use pc-relative displacements for jumps and calls)
}
function CalcDisplacement(Source, Target: pointer): integer;
begin
  Result := integer(Target) - (integer(Source) + SizeOf(integer));
end;

{ Creates a generic code hook by patching "Proc" to execute "NewProc" instead.
  Before patching, "OldProc" receives a pointer to a function that can be used
  to call the old code (though not at the same location).
}
function CreateGenericCodeHook(Proc, NewProc: pointer; var OldProc: pointer): boolean;
var
  mbi: TMemoryBasicInformation;
  len, bup: integer;
  cop: PGenCodeHookStruct;
  nj: TNearJmp;
{$IFDEF ABSOLUTE_ADDRESSES}
  aj: TAbsJmp;
{$ENDIF}
  map: TCodeMap;
begin
  Result := false;
  map:= nil;
  OldProc := nil;

  // PSDK quote: "Windows Me/98/95:  You cannot use VirtualProtect on
  //   any memory region located in the shared virtual address space
  //   (from 0x80000000 through 0xBFFFFFFF)."
  //
  // -> Hooking will fail!

  // Get size of memory region
  if VirtualQuery(Proc, mbi, SizeOf(mbi)) < SizeOf(mbi) then
    exit;

  // Analyze CPU code and find amount of memory to copy
  len := integer(mbi.RegionSize) - (integer(Proc) - integer(mbi.BaseAddress));
  map := AnalyzeCpuInstructionSequence(Proc, len);
  len := LengthOfCpuInstructionSequence(Proc, Map, PATCH_SIZE);
{$IFDEF ABSOLUTE_ADDRESSES}
  if len = 0 then
    len := LengthOfCpuInstructionSequence(Proc, Map, SizeOf(TNearJmp));
{$ENDIF}
  if len = 0 then
    exit;

  // Calculate the backup size (the remaining
  // bytes will be filled with NOPs later)
  bup := 0;
  while bup < PATCH_SIZE do
  begin
    if (Map[bup] and AISQ_SIZE_MASK) = 0 then
      inc(bup)
    else
      inc(bup, Map[bup] and AISQ_SIZE_MASK);
  end;

  // Get new memory
  cop := AllocCodeMem(SizeOf(TGenCodeHookStruct) + len);
  try
    // Initialize the header
    cop^.Header.Location := Proc;
    cop^.Header.Signature := SIGNATURE;
{$IFDEF ABS_JUMP_INDIRECT}
    cop^.Header.JumpHere := NewProc;
{$ENDIF}

    // Make backup copy
    cop^.Header.BupLen := bup;
    Move(Proc^, cop^.Header.Backup[0], bup);

    // Copy instructions (and do relocations)
    CopyCpuInstructionSequence(Proc, len, Map, @cop^.Code[0]);

    // place jump to old procedure at the end
    nj.Insn := $E9;    // call disp32
    nj.Disp := CalcDisplacement(@cop^.Code[len + 1], pinc(Proc, len));
    Move(nj, cop^.Code[len], SizeOf(TNearJmp));

    // Pointer to old procedure
    OldProc := @cop^.Code[0];

    // Finally, patch the target memory
{$IFDEF ABSOLUTE_ADDRESSES}
    if len = SizeOf(TNearJmp) then
    begin
    //nj.Insn := $E9;    // jmp disp32
      nj.Disp := CalcDisplacement(pinc(Proc), NewProc);
      Result := PatchMemory(Proc, nj, SizeOf(TNearJmp), bup);
    end
    else
    begin
{$IFDEF ABS_JUMP_INDIRECT}
      aj.Insn1 := $FF;   // jmp dword ptr [offs]
      aj.Insn2 := $25;   // ...
      aj.Addr := @cop^.Header.JumpHere;
{$ELSE}
      aj.Insn1 := $68;   // push imm32
      aj.Addr := NewProc;
      aj.Insn2 := $C3;   // ret
{$ENDIF}
      Result := PatchMemory(Proc, aj, SizeOf(TAbsJmp), bup);
    end;
{$ELSE}
  //nj.Insn := $E9;
    nj.Disp := CalcDisplacement(pinc(Proc), NewProc);
    Result := PatchMemory(Proc, nj, SizeOf(TNearJmp), bup);
{$ENDIF}

    if not Result then
    begin
      OldProc := nil;
      FreeCodeMem(cop);
    end;
  except
    OldProc := nil;
    FreeCodeMem(cop);
    raise;
  end;
end;

{ Restore a patched location and free the memory
}
procedure RemoveGenericCodeHook(var OldProc: pointer);
var
  cop: PGenCodeHookStruct;
begin
  if OldProc = nil then
    exit;

  cop := pinc(OldProc, -SizeOf(TGenCodeHookHeader));
  if (cop^.Header.Signature <> SIGNATURE) then
    exit;

  PatchMemory(cop^.Header.Location, cop^.Header.Backup[0], cop^.Header.BupLen);
  OldProc := cop^.Header.Location;
  cop^.Header.Signature := 0;

  FreeCodeMem(cop);
end;

end.
